### PR TITLE
Add --context flag

### PR DESCRIPTION
**Pull request by maruina**
Fix https://github.com/kubernetes-sigs/e2e-framework/issues/187

@vladimirvivien I know this is not what you suggested, but I spent some time looking at the cli-runtime and couldn't figure out how to wire it.
Since this is blocking our adoption and it is a pretty simple change, I'd like to ask you if we can still merge it and then I can try to take again a look at the cli-runtime fix

### CODE REVIEW

File: pkg/envconf/config.go [hunk 1]
```diff
@@ -44,6 +44,7 @@ type Config struct {
 	dryRun                  bool
 	failFast                bool
 	disableGracefulTeardown bool
+	context                 string
```
Comment (vladimirvivien) [inline pkg/envconf/config.go:47]: To avoid confusion with Go context, I would suggest naming this `k8sContext`

File: pkg/flags/flags.go [hunk 2]
```diff
@@ -175,6 +181,11 @@ func (f *EnvFlags) DisableGracefulTeardown() bool {
 	return f.disableGracefulTeardown
 }
 
+// Context returns an optional kubeconfig context to use
+func (f *EnvFlags) Context() string {
```
Comment (vladimirvivien) [inline pkg/flags/flags.go:185]: Same here, I would suggest naming the method `KubernetesContext`

**Review by vladimirvivien — COMMENTED** [review]
@maruina Hi. Thanks for the contribution. This is a good start.
I left some comments about the `context` name. Also, I don't see where in the code you are passing the kubernetes context to the the client.  I imagine that is what you want, correct ?

File: pkg/envconf/config.go (hunk 1 above)
Comment (maruina) [reply pkg/envconf/config.go:47]: We have a `kubeconfig`, what do you think about `kubecontext`? I am asking so we have only one way to prefix kubernetes things

**Comment by maruina** [comment]
> Also, I don't see where in the code you are passing the kubernetes context to the the client. I imagine that is what you want, correct ?

No :P
You are right and having a better look at the code, correct me where I'm wrong:
- we build the client with [ NewClient()](https://github.com/kubernetes-sigs/e2e-framework/blob/33c3d5efc189c44ff7dbc283a361481a264c8005/pkg/envconf/config.go#L111) where we pass the kubeconfig we got from `NewFromFlags`
- Then we call [`NewWithKubeConfigFile`](https://github.com/kubernetes-sigs/e2e-framework/blob/33c3d5efc189c44ff7dbc283a361481a264c8005/klient/client.go#L59) which is calling `conf.New(filePath)`
- If I look at the `klient/conf/config.go` I see we already have a [NewWithContextName](https://github.com/kubernetes-sigs/e2e-framework/blob/33c3d5efc189c44ff7dbc283a361481a264c8005/klient/conf/config.go#L47), but it seems we don't use it anywhere

What I could do is:
- check if`c.kubecontext` is empty or not in `NewClient()`
- call something like `NewWithKubeConfigFileAndContext()` which then calls `NewWithContextName`

It sounds a little convoluted so I'm not 100% how would you like to do it or if you have a better suggestion

**Comment by vladimirvivien** [comment]
@maruina oh yes, you are right. There is already a method that can apply the Kubernetes cluster context at test time.

So, my next question to you, do you have to pass your cluster context name as a CLI argument ? You could use an environment variable (i.e. KUBE_CONTEXT) and read it from your test, then apply that method. That way you won't even have to wait for PR or a code change.

**Comment by vladimirvivien** [comment]
@maruina I went and look at the code again. If you definitely want to move with this, introduce a `--context` as a flag, there are some additional steps you will need for it to work. I think you are almost there, but here is the complete path of changes you need:

* Add/define flag in pkg/flags - done
* Update the [envconf.Config](https://github.com/kubernetes-sigs/e2e-framework/blob/33c3d5efc189c44ff7dbc283a361481a264c8005/pkg/envconf/config.go#L33) with the context config (i.e. `kubeContext` field).
* Next update [envconf.NewFromFlags](https://github.com/kubernetes-sigs/e2e-framework/blob/33c3d5efc189c44ff7dbc283a361481a264c8005/pkg/envconf/config.go#L62:6) so that the environment can be created with the new flag
* Next, add a pair of accessor methods `Config.WithKubernetesContext` and `Config.KubernetesContext` similar to how it is done for Namespace [here](https://github.com/kubernetes-sigs/e2e-framework/blob/33c3d5efc189c44ff7dbc283a361481a264c8005/pkg/envconf/config.go#L144)
* Update [`Config.NewClient()`](https://github.com/kubernetes-sigs/e2e-framework/blob/11aa681198560751f4694b0ae613d580bbb50012/pkg/envconf/config.go#L111) to create the klient.Client with the cluster name, when the `kubeConfig` is provided.

It sounds complicated, but we are here to assist you in getting this in.

cc @harshanarayana

**Comment by maruina** [comment]
> So, my next question to you, do you have to pass your cluster context name as a CLI argument ? You could use an environment variable (i.e. KUBE_CONTEXT) and read it from your test, then apply that method. That way you won't even have to wait for PR or a code change.

Correct me if I'm wrong, but in the test we usually call `cfg.NewClient()` where cfg is a `*envconf.Config`. So I would still need to make some changes, or I'm not 100% sure what you mean with `apply that method`

Regarding the flag, thank you very much for your input, I will follow up with your suggestions

**Comment by maruina** [comment]
@vladimirvivien I think that I followed your suggestions up to this point
>Update [Config.NewClient()](https://github.com/kubernetes-sigs/e2e-framework/blob/11aa681198560751f4694b0ae613d580bbb50012/pkg/envconf/config.go#L111) to create the klient.Client with the cluster name, when the kubeConfig is provided.

When I look at the code, we call [client, err := klient.NewWithKubeConfigFile(c.kubeconfig)](https://github.com/kubernetes-sigs/e2e-framework/blob/11aa681198560751f4694b0ae613d580bbb50012/pkg/envconf/config.go#L116) to create a client with a specific kubeconfig, so my understanding is that I need to provide a new method or change `NewWithKubeConfigFile`. 

Do you want me to check if `c.kubecontext` is not an empty string inside [NewClient()](https://github.com/kubernetes-sigs/e2e-framework/blob/11aa681198560751f4694b0ae613d580bbb50012/pkg/envconf/config.go#L111)? The check we do for the kubeconfig is later in the code, when we call [conf.New(filePath)](https://github.com/kubernetes-sigs/e2e-framework/blob/11aa681198560751f4694b0ae613d580bbb50012/klient/conf/config.go#L35)

So I can call [NewWithContextName(fileName, context string)](https://github.com/kubernetes-sigs/e2e-framework/blob/11aa681198560751f4694b0ae613d580bbb50012/klient/conf/config.go#L47) but I probably need something like 
```
client, err := klient.NewWithKubeConfigFileAndKubeContext(c.kubeconfig, c.kubecontext)
```
but again, it feels a little too convoluted. What do you think?

**Comment by vladimirvivien** [comment]
> So I can call [NewWithContextName(fileName, context string)](https://github.com/kubernetes-sigs/e2e-framework/blob/11aa681198560751f4694b0ae613d580bbb50012/klient/conf/config.go#L47) but I probably need something like
> 
> ```
> client, err := klient.NewWithKubeConfigFileAndKubeContext(c.kubeconfig, c.kubecontext)
> ```

@maruina I think you are getting closer. In `Config.NewClient` you could check to see if the context is non-empty then create the client as shown below. You can  create a new constructor function `klient.NewWithContextName(file, string)` that uses `klient/conf.NewWithContextName()` to build the `klient.Client`

```go
func (c *Config) NewClient() (klient.Client, error) {
	if c.client != nil {
		return c.client, nil
	}

        if c.kubecontext != "" {
            client, err := klient.NewWithContextName(c.kubeconfig, c.kubeContext)
	    if err != nil {
		return nil, fmt.Errorf("envconfig: client failed: %w", err)
	    } 
           c.client = client
        }else{
	   client, err := klient.NewWithKubeConfigFile(c.kubeconfig)
	   if err != nil {
  	       return nil, fmt.Errorf("envconfig: client failed: %w", err)
	   }
	   c.client = client
       }

	return c.client, nil
}
```

**Comment by harshanarayana** [comment]
@vladimirvivien could we not just offload all calls from `NewClient` to the `klient.NewWithContextNam` and handle the cases there where the context is set of not? Since that is really a `klient` related case than the `Config` is it not ?

**Comment by harshanarayana** [comment]
@maruina

Sorry in advance for the really lengthy comment. 


If you check the args generated for the binaries of e2e-framework, it already has a `-kube-context` arg. May be we should consider reusing that instead of adding a new one?

```bash
❯ ./flags.test --help 2>&1 | grep -A2 kube-cont
  -kube-context string
    	Used to provide a custom KubeContext override for the file passed to the --kubeconfig
  -kubeconfig string
```

https://github.com/kubernetes-sigs/e2e-framework/blob/cd133e286ac7059c7743561cb6441542d60674f4/klient/conf/config.go#L112-L122

I think we should reuse this and modify the workflow as bit in terms of how the client is created. 

```golang
func New(fileName string) (*rest.Config, error) {
	// if filename is not provided assume in-cluster-config
	if fileName == "" {
		return rest.InClusterConfig()
	}

	if kubeContext := ResolveClusterContext(); kubeContext != "" {
		return NewWithContextName(fileName, kubeContext)
	}
	
	// create the config object from k8s config path
	return clientcmd.NewNonInteractiveDeferredLoadingClientConfig(
		&clientcmd.ClientConfigLoadingRules{ExplicitPath: fileName}, &clientcmd.ConfigOverrides{}).ClientConfig()
}
```

This should keep the code change to really minimal ? cc @vladimirvivien  




If we are to add a new one, I think here are a few suggestion. 

```golang
func NewWithKubeConfigFileAndContext(kubeConfig string, kubeContext string) (Client, error) {
	if kubeContext == "" {
		return NewWithKubeConfigFile(kubeConfig)
	}
	cfg, err := conf.NewWithContextName(kubeConfig, kubeContext)
	if err != nil {
		return nil, err
	}
	return New(cfg)
}
```
I think we can add a function like so on the `pkg/klient/client.go` and call that from `pkg/envconf/config.go` like following

```golang
func (c *Config) NewClient() (klient.Client, error) {
	if c.client != nil {
		return c.client, nil
	}

	client, err := klient.NewWithKubeConfigFileAndContext(c.kubeconfig, c.kubeContext)
	if err != nil {
		return nil, fmt.Errorf("envconfig: client failed: %w", err)
	}
	c.client = client

	return c.client, nil
}
```
This will keep the isolation intact to a good extend. 

Also a few other things to consider, 

1. If the `--kubeconfig` arg is not provided but a `--kube-context` is provided, then we should probably throw an error ?
2. I think it is better to name the arg `--kube-context` instead of `--kubecontext` for readability. 
3. Using proper camel cases in the code. Please consider using `kubeContext` instead of `kubecontext`
4. Since we are looking into it, I think the signature of `func NewWithContextName(fileName, context string) (*rest.Config, error) ` might be better changed a bit on the variable name ? The second one reads `context` which is really conflicting with the default package. Might be better to rename this `kubeContext` ?

**Comment by maruina** [comment]
I still pushed the change so if we are happy with that, we can test it :)
I can always revert to the previous state

**Comment by harshanarayana** [comment]
> @harshanarayana from which branch did you try the flags? I'm trying with the latest main but it seems I can't see that `kube-context` flag


Oh no! I think I made a mistake. I'm terribly sorry about that.. I was on your PR branch and had modified the arg name.. I totally forgot about that when I ran the help command.. That's how the variable showed up there... You still have to add that arg yourself..

**Comment by maruina** [comment]
@harshanarayana no worries. I pushed again the flag, and I was able to target a specific cluster using 
```
go test . -v -args --kubeconfig ~/.kube/config --labels category=mylabel --kube-context mycontext
```

File: pkg/flags/flags.go [hunk 3]
```diff
@@ -38,6 +38,7 @@ const (
 	flagDryRunName              = "dry-run"
 	flagFailFast                = "fail-fast"
 	flagDisableGracefulTeardown = "disable-graceful-teardown"
+	flagKubeContext             = "kube-context"
```
Comment (harshanarayana) [inline pkg/flags/flags.go:41]: I think it would be nice to have a document added in the examples indicating how to use this flag.

File: pkg/flags/flags.go [hunk 4]
```diff
@@ -247,6 +259,10 @@ func ParseArgs(args []string) (*EnvFlags, error) {
 		flag.BoolVar(&disableGracefulTeardown, disableGracefulTeardownFlag.Name, false, disableGracefulTeardownFlag.Usage)
 	}
 
+	if flag.Lookup(kubeContextFlag.Name) == nil {
```
Comment (harshanarayana) [inline pkg/flags/flags.go:262]: I think this should be checked against the presence of the `-kubeconfig` argument as well.  Without that, this argument doesn't really hold much of a values. Is it not ?

**Review by harshanarayana — CHANGES_REQUESTED** [review]

**Comment by harshanarayana** [comment]
@maruina  Leaving a few minor comments. Once done, let us see if @vladimirvivien  is okay with this approach. 

/cc @ShwethaKumbla

File: pkg/flags/flags.go (hunk 4 above)
Comment (maruina) [reply pkg/flags/flags.go:262]: I thought we were using the default location at `~/.kube/config`, but I'll add a check

**Review by vladimirvivien — COMMENTED** [review]
Apologies @maruina and @harshanarayana for the delay in response. Thank you @harshanarayana for the thoughtful feedback in this PR. I think it is ready. @maruina please squash the commits and I will take one last look to get this in.

**Comment by maruina** [comment]
@vladimirvivien squash done, please ping me if there is anything else I can do

File: pkg/flags/flags.go [hunk 5]
```diff
@@ -264,6 +280,10 @@ func ParseArgs(args []string) (*EnvFlags, error) {
 		panic(fmt.Errorf("--fail-fast and --parallel are mutually exclusive options"))
 	}
 
+	if kubeconfig == "" && kubeContext != "" {
+		panic(fmt.Errorf("--kube-context requires the --kubeconfig option"))
```
Comment (vladimirvivien) [inline pkg/flags/flags.go:284]: While this is true, it is not necessary to have the --kubeconfig flag for this to work. This is because if [`ResolveKubeConfigFile`](https://github.com/kubernetes-sigs/e2e-framework/blob/db4e9fe7e23d2602c34d1fe3853625e754b1fa41/klient/conf/config.go#L67) is called during the construction of the client instance, it will attempt to resolve the kubeconfig automatically.  

Maybe that check should be moved [above](https://github.com/kubernetes-sigs/e2e-framework/pull/197/files#diff-3923bb7d5d40d33c81ef1c40584ce3ef3f81c7c6bb1ffb2adfc842b745533a7cR41) when setting the flags up.

File: klient/conf/config.go [hunk 6]
```diff
@@ -38,6 +38,11 @@ func New(fileName string) (*rest.Config, error) {
 		return rest.InClusterConfig()
 	}
 
+	// set the desired kube context
+	if kubeContext := ResolveClusterContext(); kubeContext != "" {
```
Comment (vladimirvivien) [inline klient/conf/config.go:42]: This check should also check to see if kubeConfig is available (either via the --kubeconfig flag value or via [`ResolveKubeConfigFile`](https://github.com/kubernetes-sigs/e2e-framework/blob/db4e9fe7e23d2602c34d1fe3853625e754b1fa41/klient/conf/config.go#L67). If the value of KubeConfig is not provided, this should error out. See comment below for detail.

**Review by vladimirvivien — COMMENTED** [review]
@maruina I have left one comment (split in two places) for one last check. Then I think it's ready for a merge.

File: klient/conf/config.go (hunk 6 above)
Comment (maruina) [reply klient/conf/config.go:42]: > If the value of KubeConfig is not provided, this should error out.

But if we are not passing a kubeconfig we should hit https://github.com/kubernetes-sigs/e2e-framework/blob/db4e9fe7e23d2602c34d1fe3853625e754b1fa41/klient/conf/config.go#L37 because we're assuming in-cluster-config

My understanding is that we call `ResolveKubeConfigFile` in https://github.com/kubernetes-sigs/e2e-framework/blob/db4e9fe7e23d2602c34d1fe3853625e754b1fa41/examples/real_cluster/main_test.go#L38 and then we construct the config from that file, without using any flag.

I think that you're saying that we could use `ResolveKubeConfigFile` to automatically get the kubeconfig *and* we could use the flags to also do additional configurations, am I correct? But `NewFromFlags()` doesn't take any argument, so I think I'm missing something

Comment (vladimirvivien) [reply klient/conf/config.go:42]: Hi @maruina . You are correct in your last thinking. 
We could force --kubeconfig to be provided when --context is used or (as i tried to suggest)
* Check for value of kubeConfig
* if empty, call ResolveKubeConfigFile
* if still empty, then error out

Also, if there is a better place to make that check work, you can put it where it works best.

File: examples/flags/README.md [hunk 7]
```diff
@@ -28,6 +28,7 @@ There are several supported flags (for more accurate list, see package `pkg/flag
 * `features`
 * `labels`
 * `kubeconfig`
+* `kube-context`
```
Comment (vladimirvivien) [inline examples/flags/README.md:31]: @maruina one more minor suggestion
Let's call the flag just `context`. This is to stay aligned with the naming convention (usually) followed by other tools in the kubernetes ecosystem.

**Review by vladimirvivien — COMMENTED** [review]
This is taking shape nicely. Thanks for all the work.

File: examples/flags/README.md (hunk 7 above)
Comment (maruina) [reply examples/flags/README.md:31]: Yep, done. That's why I was suggesting it in the fist place

File: klient/conf/config.go (hunk 6 above)
Comment (maruina) [reply klient/conf/config.go]: Ok, I think I've done that. Please have a look :)

File: klient/conf/config.go [hunk 8]
```diff
@@ -33,14 +34,33 @@ var DefaultClusterContext = ""
...
+		resolvedKubeConfigFile = fileName
+	}
+
+	// if resolvedKubeConfigFile is still empty, assume in-cluster config
+	if resolvedKubeConfigFile == "" {
+		if context == "" {
+			return rest.InClusterConfig()
+		}
+		// if in-cluster can't use the --context flag
+		return nil, errors.New("cannot use --context flag without a kubeconfig file")
```
Comment (vladimirvivien) [inline klient/conf/config.go:53]: I would reword the error to just mention "cluster context" instead of "--context flag" since context can come in programmatically at this point in the code, not necessarily via flag.

**Review by vladimirvivien — COMMENTED** [review]
@maruina I only have one nit comment. After this I think it's ready.

File: klient/conf/config.go [hunk 9]
```diff
@@ -33,14 +34,33 @@ var DefaultClusterContext = ""
 // New returns Kubernetes configuration value of type *rest.Config.
 // filename is kubeconfig file
 func New(fileName string) (*rest.Config, error) {
-	// if filename is not provided assume in-cluster-config
+	var resolvedKubeConfigFile string
+	context := ResolveClusterContext()
```
Comment (harshanarayana) [inline klient/conf/config.go:38]: Please do not shadow the context here with the default package name. Use `kubeContext` instead

File: klient/conf/config.go (hunk 8 above)
Comment (harshanarayana) [reply klient/conf/config.go:53]: The argument is `--kube-context` right? Can you please update the name of the arg as well

File: klient/conf/config.go (hunk 9 above)
Comment (maruina) [reply klient/conf/config.go:38]: Ops you are right, sorry. Fixed 🤦

File: klient/conf/config.go (hunk 8 above)
Comment (maruina) [reply klient/conf/config.go:53]: @harshanarayana , no as @vladimirvivien said in the previous comment, we switched back to `--context` to be consistent with other Kubernetes CLI. For example `kubectl` uses `--context`.

Comment (harshanarayana) [reply klient/conf/config.go]: Sorry, I totally missed that part of the discussion. @maruina Thanks

File: pkg/flags/flags.go [hunk 10]
```diff
@@ -277,6 +293,7 @@ func ParseArgs(args []string) (*EnvFlags, error) {
 		dryRun:                  dryRun,
 		failFast:                failFast,
 		disableGracefulTeardown: disableGracefulTeardown,
+		context:                 context,
```
Comment (harshanarayana) [inline pkg/flags/flags.go:296]: @vladimirvivien I know you dissed about keep this consistent with the kubectl in terms of args. Which Is perfectly fine. But I think we should use the internal variables/attributes in such a way that it won't conflict with the `context` package. That can lead to a lot of confusion. Do you think it will be ok to at least name the internal attributes/variables as `kubeContext` but retain the flag argument as `--context` ?

**Review by harshanarayana — COMMENTED** [review]
Sorry, for doing this in bits. Leaving a couple of more nitpicks. Missed them last time as I was on the phone.

File: pkg/flags/flags.go (hunk 10 above)
Comment (maruina) [reply pkg/flags/flags.go:296]: Not @vladimirvivien but I think you have a point and I can change that. I didn't think about the `context` package and the shadowing at all when writing this.

Comment (vladimirvivien) [reply pkg/flags/flags.go:296]: @maruina @harshanarayana Yes, I thought about `context` causing confusion with context package. So, what I am suggesting is just the name of the flag being called `context`. However internally, the variable name should be kubeContext (everywhere it's used).  Hope that makes sense.

Comment (vladimirvivien) [reply pkg/flags/flags.go:296]: I think after that, this will be ready to go and we can have a release soon.

File: pkg/envconf/config.go [hunk 11]
```diff
@@ -44,6 +44,7 @@ type Config struct {
 	dryRun                  bool
 	failFast                bool
 	disableGracefulTeardown bool
+	kubeContext             string
```
Comment (maruina) [inline pkg/envconf/config.go:47]: @vladimirvivien and @harshanarayana, even if we're parsing directly from flags in `ResolveClusterContext()`, I thought about adding the flag to the `Config` object for consistency.

If you think it's not necessary, I'm happy to drop this change

**Comment by maruina** [comment]
I tested the squashed commit locally and I can run some tests to a target context with
```
❯ go test . -v -args --labels category=my-category -context my-context
```
The nice thing is that I don't have to specify the kubeconfig flag anymore.

If we pass an invalid file, for example
```
❯ go test . -v -args --labels category=my-category -context my-context --kubeconfig /tmp/foo
F0214 22:56:20.312484   71123 env.go:369] Setup failure: create namespace func: envconfig: client failed: cannot use a cluster context without a valid kubeconfig file
FAIL	github.com/my-repo/e2eframework	0.362s
FAIL
```
That is because we bubble up `kubeconfig` as an empty string and we don't get the associated error - in this case the file doesn't exist.

File: pkg/flags/flags.go (hunk 10 above)
Comment (harshanarayana) [reply pkg/flags/flags.go]: @vladimirvivien  @maruina  Works for me.

**Review by vladimirvivien — COMMENTED** [review]
I apologize, I meant to send this earlier this week. After this, I think it's ready for a merge.

**Comment by maruina** [comment]
@vladimirvivien updated the code :)

**Review by vladimirvivien — APPROVED** [review]
Thanks @maruina for this awesome contribution.

**Comment by harshanarayana** [comment]
@maruina 

```bash
go: downloading github.com/josharian/intern v1.0.0
E0217 13:49:12.378771    9620 env.go:352] Recovering from panic and running finish actionsenvconfig: client failed: open /var/run/secrets/kubernetes.io/serviceaccount/token: no such file or directory
program not built with -cover
ok  	sigs.k8s.io/e2e-framework/examples/crds	41.446s
=== RUN   TestListPods
=== RUN   TestListPods/pod_list
=== RUN   TestListPods/pod_list/pods_from_kube-system
    k8s_test.go:33: envconfig: client failed: open /var/run/secrets/kubernetes.io/serviceaccount/token: no such file or directory
--- FAIL: TestListPods (0.00s)
    --- FAIL: TestListPods/pod_list (0.00s)
        --- FAIL: TestListPods/pod_list/pods_from_kube-system (0.00s)
FAIL
coverage: [no statements]
FAIL	sigs.k8s.io/e2e-framework/examples/custom_env_funcs	38.357s
E0217 13:50:14.534158   15810 env.go:352] Recovering from panic and running finish actionsenvconfig: client failed: open /var/run/secrets/kubernetes.io/serviceaccount/token: no such file or directory
program not built with -cover
ok  	sigs.k8s.io/e2e-framework/examples/decoder	20.[72](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_e2e-framework/197/pull-e2e-framework-test/1626578895165722624#1:build-log.txt%3A72)6s
```

Seems like new tests are breaking the tests

File: klient/conf/config.go [hunk 12]
```diff
@@ -33,14 +34,34 @@ var DefaultClusterContext = ""
...
+	kubeContext := ResolveClusterContext()
+
+	// resolve the kubeconfig file
 	if fileName == "" {
-		return rest.InClusterConfig()
+		resolvedKubeConfigFile = ResolveKubeConfigFile()
+		if fileName != "" {
+			resolvedKubeConfigFile = fileName
+		}
+	}
```
Comment (harshanarayana) [inline klient/conf/config.go:46]: Sorry, I missed this one 🤦 

This can't be happening right ? Your first check is `fileName == ""` and inside the lop you have a `fileName != ""` without first setting the `fileName` to a new values.

**Comment by maruina** [comment]
@vladimirvivien I think that I actually broke something, let me see where the problem is because I think I tested everything before adding the two new methods so I'm not sure how I broke things

**Review by harshanarayana — CHANGES_REQUESTED** [review]

File: klient/conf/config.go (hunk 12 above)
Comment (harshanarayana) [reply klient/conf/config.go:46]: @maruina  I think I know why. Check the above one. 

```golang
// resolve the kubeconfig file
	resolvedKubeConfigFile = fileName
	if fileName == "" {
		resolvedKubeConfigFile = ResolveKubeConfigFile()
	}
```

Should be like above.

Comment (maruina) [reply klient/conf/config.go]: Yes, of course 🤦 
Tests are passing locally now.

**Comment by vladimirvivien** [comment]
@maruina ping me when you think you got it fixed.

**Comment by maruina** [comment]
@vladimirvivien all the tests are passing locally, can you please retest?

**Review by harshanarayana — APPROVED** [review]
;lgtm

**Comment by vladimirvivien** [comment]
Thanks @maruina !!

**Comment by maruina** [comment]
Thanks both of you for your patience and help
